### PR TITLE
feat(bigquery): add support for snapshot/restore

### PR DESCRIPTION
**Pull request by shollyman**
Secondary changes:  This complicated some internal wrapping functions
for TableMetadata.  The fact that TableMetadata can now contain a Table
reference for other tables means we need to pass around an additional
instance of the Client, as user may want to do something like call
Metadata() on the embedded reference to the base table in a
snapshot definition.

### CODE REVIEW

**Review by stephaniewang526 — APPROVED** [review]

File: bigquery/integration_test.go [hunk 1]
```diff
@@ -438,6 +438,96 @@ func TestIntegration_TableMetadata(t *testing.T) {
...
+	}
+	if meta.Type != Snapshot {
+		t.Errorf("expected snapshot table type, got %s", meta.Type)
+	}
+	want := &SnapshotDefinition{
+		BaseTableReference: dataset.Table(baseTableID),
+		SnapshotTime:       targetTime,
+	}
+	if diff := testutil.Diff(meta.SnapshotDefinition, want, cmp.AllowUnexported(Table{}), cmpopts.IgnoreUnexported(Client{}), cmpopts.EquateApproxTime(time.Millisecond)); diff != "" {
+		t.Fatalf("got=-, want=+:\n%s", diff)
```
Comment (tbpg) [inline bigquery/integration_test.go:498]: Should generally include what failed in the error message.

https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures

File: bigquery/integration_test.go [hunk 2]
```diff
@@ -438,6 +438,96 @@ func TestIntegration_TableMetadata(t *testing.T) {
...
+	}
+	status, err = job.Wait(ctx)
+	if err != nil {
+		t.Fatalf("polling restore failed: %v", err)
+	}
+	if status.Err() != nil {
+		t.Fatalf("restore failed in error: %v", status.Err())
+	}
+
+	meta2, err := dataset.Table(restoreID).Metadata(ctx)
```
Comment (tbpg) [inline bigquery/integration_test.go:517]: Consider `restoredMeta` or something more descriptive. At first, I thought we could just reuse `meta`.

File: bigquery/table.go [hunk 3]
```diff
@@ -223,6 +230,40 @@ func bqToMaterializedViewDefinition(q *bq.MaterializedViewDefinition) *Materiali
...
+}
+
+func bqToSnapshotDefinition(q *bq.SnapshotDefinition, c *Client) *SnapshotDefinition {
+	if q == nil {
+		return nil
+	}
+	sd := &SnapshotDefinition{
+		BaseTableReference: bqToTable(q.BaseTableReference, c),
+	}
+	if t, err := time.Parse(time.RFC3339, q.SnapshotTime); err == nil {
```
Comment (tbpg) [inline bigquery/table.go:261]: Is it correct to ignore this error? Add a comment?

File: bigquery/integration_test.go (hunk 1 above)
Comment (shollyman) [reply bigquery/integration_test.go]: Done.

File: bigquery/integration_test.go (hunk 2 above)
Comment (shollyman) [reply bigquery/integration_test.go]: Done.

File: bigquery/table.go (hunk 3 above)
Comment (shollyman) [reply bigquery/table.go:263]: Added an explanatory comment.

**Review by tbpg — APPROVED** [review]

**Review by codyoss — APPROVED** [review]
LGTM

**Pull request merged**